### ISSUE TITLE
special case when iterating a list in reverse

### ISSUE THREAD

**Issue by lewis6991** [EmmyLuaLs/emmylua-analyzer-rust#521]
Feel free to close this if it is too complex or difficult.

Now array elements are nilable (which I think was a good decision), I'm getting a few errors when iterating lists in reverse.

![Image](https://github.com/user-attachments/assets/d8d5e02a-6fb8-4af6-926d-c747465b80af)

```lua
--- @param _x string
local function foo(_x) end

local list = {} --- @type string[]

for i = #list, 1, -1 do
  foo(list[i])
end
```

Would it be worth trying to elide this warning if the following conditions hold.

- Loop statement must be of `for <i> = #<list>, <bottom>, -1`
- `<i>` and `<list>` are never mutated in the loop body
- `<list>` is not global
- `<list>` is not passed to any functions that could mutate it

If these hold then `<list>[<i>]` should never be `nil`.

If this is too complex then I'll accept having to do any of:

```lua
for i = #list, 1, -1 do
  foo(assert(list[i]))
end

for i = #list, 1, -1 do
  local l = assert(list[i])
  foo(l)
end

for i = #list, 1, -1 do
  local l = list[i]
  --- @cast l -?
  foo(l)
end

for i = #list, 1, -1 do
  local l = list[i]
  if l then
    foo(l)
  end
end

-- anything better?
```

**Comment by xuhuanzy** [comment]
Complete control flow analysis needs to be implemented. Initially, cppcxy found it troublesome and didn't implement it, but in reality, it's essential. All the major issues we're facing now stem from this.

**Comment by lewis6991** [comment]
This now works:

```lua
--- @param _x string
local function foo(_x) end

local list = {} --- @type string[]

for i = 1, #list do
  foo(list[i])
end
```

Shouldn't be too difficult to special case:

```lua
for i = #list, 1, -1 do
  foo(list[i])
end
```
